implement torch.addr using TensorIterator based kernels

RockingJavaBean opened the pull request:
Fixes #47313
Related #44364

This PR implements `torch.addr` function using `TensorIterator` with `cpu_kernel_vec` and `gpu_kernel`.
It helps reduce memory usage, improve performance, and fix the bug when `beta` or `alpha` is a complex number.

Todo
- [x] benchmarking `torch.addr` for the change of this PR, as well as the legacy TH implementation used in PyTorch 1.6.

RockingJavaBean committed:
implement torch.addr using TensorIterator based kernels

RockingJavaBean commented:
@JackCaoG This PR implements `torch.addr` using `TensorIterator` with corresponding cpu and cuda kernels.
The `test_broadcast_fused_matmul_xla` failed in the XLA CI job.
```
Nov 10 15:49:31 ERROR [3.464s]: test_broadcast_fused_matmul_xla (__main__.TestTorchDeviceTypeXLA)
Nov 10 15:49:31 ----------------------------------------------------------------------
Nov 10 15:49:31 Traceback (most recent call last):
Nov 10 15:49:31   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
Nov 10 15:49:31     result = test_fn(self, *args)
Nov 10 15:49:31   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 8451, in test_broadcast_fused_matmul
Nov 10 15:49:31     r0 = fntorch(t0_small, t1, t2)
Nov 10 15:49:31 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
```

Please kindly confirm whether I could skip the `torch.addr` XLA error with the below updates for now.
- remove the `addr` from the `fns` list
- add a new test (`test_broadcast_fused_addr`) with `onlyOnCPUAndCUDA` annotation

https://github.com/pytorch/pytorch/blob/ae5c2febb912066f1a8dec8b54451c09195b2c6d/test/test_torch.py#L8419-L8423

RockingJavaBean commented:
The performances of `torch.addr` were benchmarked for the following versions of PyTorch.
- This PR: TensorIterator based kernels implementation
- PyTorch 1.8 nightly build: composite ops based implementation
- Pytorch 1.6.0: legacy TH code, implemented using Blas solutions

As `torch.utils.benchmark` were not introduced in PyTorch 1.6.0, I manually ported it by copying the corresponding codes to the installed path of PyTorch 1.6.0.

<details>
<summary>Benchmark script</summary>
<p>

```python
import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
from itertools import product

results = []
num_threads = 1

size = 5000

beta = 1.2
alpha = 2.2

dtypes = torch.testing.get_all_fp_dtypes()
devices = [torch.device('cpu'), torch.device('cuda')]

for device, dtype in product(devices, dtypes):
    if torch.__version__ == '1.6.0' and device == torch.device('cuda') and dtype == torch.bfloat16:
        continue
    if torch.__version__ == '1.6.0' and device == torch.device('cpu') and dtype == torch.half:
        continue

    M = torch.randn(size, size).to(device=device, dtype=dtype)
    vec1 = torch.randn(size).to(device=device, dtype=dtype)
    vec2 = torch.randn(size).to(device=device, dtype=dtype)

    tasks = [("torch.addr(M, vec1, vec2, beta=beta, alpha=alpha)", f"{dtype}")]
    timers = [Timer(stmt=stmt,
                    num_threads=num_threads,
                    label=f"{torch.__version__}, size: {size}",
                    sub_label=f"{device}",
                    description=dtype,
                    globals=globals()) for stmt, dtype in tasks]

    repeats = 3
    for i, timer in enumerate(timers * repeats):
        results.append(timer.blocked_autorange())
        print(f"\r{device}-{dtype} {i + 1} / {len(timers) * repeats}")

comparison = Compare(results)
comparison.print()

```
</p>
</details>


**This PR**
```
[-------------------------- 1.8.0a0+9cad980, size: 5000 --------------------------]
            |  torch.float32  |  torch.float64  |  torch.float16  |  torch.bfloat16
1 threads: ------------------------------------------------------------------------
      cpu   |       65.9      |      164.4      |      909.9      |      190.7
      cuda  |        4.1      |        8.2      |        3.6      |        3.9

Times are in milliseconds (ms).
```

**PyTorch 1.8 nightly build**
```
[------------------------- 1.8.0.dev20201110, size: 5000 -------------------------]
            |  torch.float32  |  torch.float64  |  torch.float16  |  torch.bfloat16
1 threads: ------------------------------------------------------------------------
      cpu   |      191.8      |      382.4      |      1256.1     |      300.1
      cuda  |       16.9      |       32.4      |         9.9     |       10.2

Times are in milliseconds (ms).
```

**PyTorch 1.6.0**
```
[------------------------------- 1.6.0, size: 5000 -------------------------------]
            |  torch.float32  |  torch.float64  |  torch.bfloat16  |  torch.float16
1 threads: ------------------------------------------------------------------------
      cpu   |       59.0      |      120.8      |      152.6       |
      cuda  |       13.2      |       24.9      |                  |       5.5

Times are in milliseconds (ms).
```
The CPU path does not support `torch.float16`, while the CUDA path does not support `torch.bfloat16` in PyTorch 1.6.0.

**Result**
The TensorIterator based implementations outperformed both the composite ops and BLAS based implementions in most cases.
It was sightly slower than the BLAS ones in terms of `torch.float32` and `torch.float64` for the CPU path, while it increased the CUDA performance to a significant percentange.

Please kindly help review this PR. @ngimel @mruberry

RockingJavaBean commented on test/test_linalg.py:
```diff
@@ -70,41 +67,84 @@ def run_test_case(m, a, b, beta=1, alpha=1):
+        # when beta is not zero
+        self._test_addr_vs_numpy(device, dtype, beta=0.5, alpha=2)
+        if dtype in torch.testing.get_all_complex_dtypes():
+            self._test_addr_vs_numpy(device, dtype, beta=(0 + 0.1j), alpha=(0.2 - 0.2j))
```
This is to test the scenarios when `beta` or `alpha` is a complex number.

ailzhang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -143,34 +147,85 @@ static void check_1d(const Tensor& t, const char* arg, const char* fn) {
-  TORCH_CHECK(canCast(result_dtype, result.scalar_type()),
-              "result type ", result_dtype,
-              " can't be cast to the desired output type ", result.scalar_type());
+Tensor& addr_out(Tensor &result,
```
One difference made in this PR is the old implementation works for all backends but now it only support CPU and CUDA only. 
So here are a few missing changes in this PR:
1. change the addr.out entry in native_functions.yaml from DefaultBackend to CPU,CUDA only.  
2. Save the old implementation of addr_out in a math_addr_out function and put it as Math keyword in native_functions.yaml. see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#choosing-the-right-dispatch-keyword for more details. 
3. Add a test in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/math_kernel_test.cpp to make sure the new kernels you wrote for CPU & CUDA produces the same results as the old kernels we have. 
The second item should fix the xla test failure properly. 
All of the steps above should be smalll changes I think. 
Please let me know if you have followup questions! Thanks!

RockingJavaBean replied:
@ailzhang Thank you so much for your kind help in fixing the XLA failures, and providing the detailed docs and samples.
I have addressed the suggested changes and the XLA CI checks passed.

ailzhang reviewed (changes requested):
Left some comments on the backend support. Please feel free to reach out if you have any questions!

RockingJavaBean committed:
math key for torch.addr

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

ngimel commented on aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp:
```diff
@@ -0,0 +1,52 @@
+     [=](scalar_t self_val,
+         scalar_t vec1_val,
+         scalar_t vec2_val) __ubsan_ignore_undefined__ -> scalar_t {
+        return beta_val * self_val + alpha_val * vec1_val * vec2_val;
```
this does not ignore `self_val` if beta=0, thus nans and infs in `self` won't be ignored, same in cuda kernel

RockingJavaBean replied:
Thanks so much for pointing this out. I have updated the `CPU` and `cuda` kernels, the corresponding CPP and Pytorch test codes have been updated as well.

ngimel commented on test/test_linalg.py:
```diff
@@ -296,41 +293,83 @@ def run_test_case(m, a, b, beta=1, alpha=1):
-        run_test_case(m_scalar, a, b)
+        check(m_scalar, a, b, beta, alpha)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
nit: numpy is required, no need for this skipIf

RockingJavaBean replied:
Addressed.

ngimel reviewed (approved):
couple small comments, also, can you please rebase?

RockingJavaBean committed:
ignore nans and infs in self when beta equals 0

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_addr_iter

RockingJavaBean committed:
fix lints

RockingJavaBean committed:
fix CI failures of cpu kernel

mruberry commented on torch/_torch_docs.py:
```diff
@@ -559,13 +559,7 @@ def merge_dicts(*dicts):
-    It can be implemented using :func:`torch.outer` as
-    ``alpha * torch.outer(vec1, vec2) + beta * input`` when :attr:`beta` is not zero,
-    and as ``alpha * torch.outer(vec1, vec2)`` when :attr:`beta` is zero.
+:attr:`alpha` must be real numbers, otherwise they should be integers.
```
This paragraph is out of date and the term "real numbers" is ambiguous.

RockingJavaBean replied:
Thanks so much for the suggestion, the paragraph has been removed.

mruberry reviewed (approved):
Overall looks good; I pointed out one docs issue.

RockingJavaBean committed:
remove out-of-date doc paragraph

RockingJavaBean committed:
fix ci issues due to gpu_kernel warnings

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_addr_iter

RockingJavaBean commented:
@mruberry @ngimel Thanks so much for the review.
I have updated the PR according to the comments and rebased the code, `Circle CI` checks seem all passed except for one broken upstream CI.
Please kindly take a look.
